File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* head1, ListNode* head2) {
        ListNode dummyHead;
        ListNode* ansCurr = &dummyHead;
        
        while(head1 != NULL && head2 != NULL) {
            if(head1->val < head2->val) {
                ansCurr->next = head1;
                head1 = head1->next;
            } else {
                ansCurr->next = head2;
                head2 = head2->next;
            }
            ansCurr = ansCurr->next;
        }
        
        // Attach the remaining elements of head1 or head2
        ansCurr->next = head1 ? head1 : head2;
        
        return dummyHead.next;
    }
};

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    if list1 == nil && list2 == nil {
        return list1
    } 
    
    if list1 == nil {
        return list2
    }
    
    if list2 == nil {
        return list1
    }
    
    c1 := list1
    c2 := list2
    var res *ListNode // assigned as nil
    
    for c1 != nil && c2 != nil {
        if c1.Val <= c2.Val {
            if res == nil {
                res = c1
            } else {
                res.Next = c1
                res = res.Next
            }
            c1 = c1.Next
        } else {
            if res == nil {
                res = c2
            } else {
                res.Next = c2
                res = res.Next
            }
            c2 = c2.Next
        }
    }
    
    // remaining list1 elements
    if c1 != nil {
        if res == nil {
            res = c1
        } else {
            res.Next = c1
        }
    }
    
    // remaining list2 elements 
    if c2 != nil {
        if res == nil {
            res = c2
        } else {
            res.Next = c2
        }
    }
    
    if list1.Val <= list2.Val {
        return list1
    } else {
        return list2
    }
}